providers/providers.go

```diff
@@ -310,6 +314,21 @@ func (pm *ProviderManager) AddProvider(ctx context.Context, k []byte, val peer.I
 	}
 }
 
+func (pm *ProviderManager) addProv(k []byte, p peer.ID) error {
+	now := time.Now()
+	// If there is something on cache, update cache
+	if cached, ok := pm.cache.Get(string(k)); ok {
+		pSet := cached.(*providerSet)
+		pSet.providers = append(pSet.providers, p)
+		pSet.set[p] = now
```

Comment:
I don't understand why `setVal` shouldn't be used? As far as I can tell, `providerSet` is tracking which peers are in `pSet.providers` via the map `pSet.set`. So if they're in the `pSet.set` they're also in `pSet.providers`.

`setVal` guards from adding the same provider twice by checking if they're in the map first.